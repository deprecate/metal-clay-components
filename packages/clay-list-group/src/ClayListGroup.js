import 'clay-checkbox';
import 'clay-dropdown';
import 'clay-sticker';

import Component from 'metal-component';
import defineWebComponent from 'metal-web-component';
import Soy from 'metal-soy';
import {Config} from 'metal-state';

import templates from './ClayListGroup.soy.js';

/**
 * Metal ClayListGroup component.
 */
class ClayListGroup extends Component {}

/**
 * State definition.
 * @static
 * @type {!Object}
 */
ClayListGroup.STATE = {
	/**
	 * CSS classes to be applied to the element.
	 * @instance
	 * @memberof ClayListGroup
	 * @type {?string|undefined}
	 * @default undefined
	 */
	elementClasses: Config.string(),

	/**
	 * Id to be applied to the element.
	 * @instance
	 * @memberof ClayListGroup
	 * @type {?string|undefined}
	 * @default undefined
	 */
	id: Config.string(),

	/**
	 * List of items.
	 * @instance
	 * @memberof ClayListGroup
	 * @type {!array}
	 * @default undefined
	 */
	items: Config.arrayOf(
		Config.shapeOf({
			actionItems: Config.arrayOf(
				Config.shapeOf({
					disabled: Config.bool().value(false),
					events: Config.object(),
					label: Config.string().required(),
					separator: Config.bool().value(false),
					url: Config.string().required(),
				})
			),
			description: Config.string(),
			icon: Config.string(),
			iconShape: Config.oneOf(['circle', 'rounded']).value('rounded'),
			selected: Config.bool().value(false),
			title: Config.any().required(),
		})
	).required(),

	/**
	 * Flag to indicate if the list group items are selectable.
	 * @instance
	 * @memberof ClayListGroup
	 * @type {?bool|undefined}
	 * @default false
	 */
	selectable: Config.bool().value(false),

	/**
	 * The path to the SVG spritemap file containing the icons.
	 * @instance
	 * @memberof ClayListGroup
	 * @type {?string|undefined}
	 * @default undefined
	 */
	spritemap: Config.string(),
};

defineWebComponent('clay-list-group', ClayListGroup);

Soy.register(ClayListGroup, templates);

export {ClayListGroup};
export default ClayListGroup;
