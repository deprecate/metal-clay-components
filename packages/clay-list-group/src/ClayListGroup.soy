{namespace ClayListGroup}

/**
 * This renders the component's whole content.
 */
{template .render}
	{@param items: list<[
		actionItems: list<[
			disabled: bool,
			events: ?,
			indicatorSymbol: string,
			label: string,
			quickAction: bool,
			separator: bool,
			url: string
		]>,
		description: string,
		icon: string,
		iconShape: string,
		label: string,
		labelStyle: string,
		selected: bool,
		title: string,
		url: string
	]>}
	{@param? elementClasses: string}
	{@param? handleItemToggled_: any}
	{@param? id: string}
	{@param? selectable: bool}
	{@param? spritemap: string}
	{@param? title: string}

	{let $attributes kind="attributes"}
		class="list-group show-quick-actions-on-hover
			{if $elementClasses}
				{sp}elementClasses
			{/if}
		"

		{if $id}
			id="{$id}"
		{/if}
	{/let}

	<ul {$attributes}>
		{if $title}
			<li class="list-group-header">
				<h3 class="list-group-header-title">{$title}</h3>
			</li>
		{/if}

		{foreach $item in $items}
			{call ClayListGroupItem.render}
				{param actionItems: $item.actionItems /}
				{param description: $item.description /}
				{param events: ['itemToggled': $handleItemToggled_ ] /}
				{param icon: $item.icon /}
				{param iconShape: $item.iconShape /}
				{param label: $item.label /}
				{param labelStyle: $item.labelStyle /}
				{param selectable: $selectable /}
				{param selected: $item.selected /}
				{param spritemap: $spritemap /}
				{param title: $item.title /}
				{param url: $item.url /}
			{/call}
		{/foreach}
	</ul>
{/template}