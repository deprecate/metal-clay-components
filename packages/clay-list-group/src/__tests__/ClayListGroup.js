import ClayListGroup from '../ClayListGroup';

const spritemap = '../node_modules/lexicon-ux/build/images/icons/icons.svg';

let component;

let actionItems = [
	{
		label: 'Option 1',
		url: '#1',
	},
	{
		label: 'Option 2',
		separator: true,
		url: '#2',
	},
	{
		label: 'Option 3',
		url: '#3',
	},
];

let actionItemsWithQuickItems = [
	{
		indicatorSymbol: 'trash',
		label: 'Option 1',
		quickAction: true,
		url: '#1',
	},
	{
		indicatorSymbol: 'download',
		label: 'Option 2',
		quickAction: true,
		separator: true,
		url: '#2',
	},
	{
		indicatorSymbol: 'info-circle-open',
		label: 'Option 3',
		quickAction: true,
		url: '#3',
	},
];

describe('ClayListGroup', function() {
	afterEach(() => {
		if (component) {
			component.dispose();
		}
	});

	it('should render the default markup', () => {
		component = new ClayListGroup({
			items: [],
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayListGroup with classes', () => {
		component = new ClayListGroup({
			elementClasses: 'my-custom-class',
			items: [],
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayListGroup with id', () => {
		component = new ClayListGroup({
			id: 'myId',
			items: [],
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a selectable ClayListGroup', () => {
		component = new ClayListGroup({
			items: [
				{
					description: 'Description 1',
					title: 'Item 1',
					url: '#1',
				},
				{
					description: 'Description 2',
					title: 'Item 2',
					url: '#2',
				},
				{
					description: 'Description 3',
					title: 'Item 3',
					url: '#3',
				},
			],
			selectable: true,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a selectable ClayListGroup with header title', () => {
		component = new ClayListGroup({
			items: [
				{
					title: 'Item 1',
				},
			],
			title: 'Group Header 1',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayListGroup and emit an event on item toggled', () => {
		const spy = jest.fn();

		component = new ClayListGroup({
			events: {itemToggled: spy},
			items: [
				{
					title: 'Item 1',
				},
			],
			selectable: true,
		});

		component.element.querySelector('input[type="checkbox"]').click();

		expect(spy).toHaveBeenCalled();
	});

	it('should fail when no items are passed', function() {
		expect(() => {
			component = new ClayListGroup();
		}).toThrow();
	});
});
