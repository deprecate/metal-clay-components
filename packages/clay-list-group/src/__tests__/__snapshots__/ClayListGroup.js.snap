// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayListGroup should render a ClayListGroup with classes 1`] = `<ul class="list-group show-quick-actions-on-hover elementClasses my-custom-class"></ul>`;

exports[`ClayListGroup should render a ClayListGroup with id 1`] = `<ul class="list-group show-quick-actions-on-hover" id="myId"></ul>`;

exports[`ClayListGroup should render a selectable ClayListGroup 1`] = `
<ul class="list-group show-quick-actions-on-hover">
  <li class="list-group-item list-group-item-flex">
    <div class="flex-col">
      <div class="custom-control custom-checkbox">
        <label>
          <input class="custom-control-input" ref="input" type="checkbox">
          <span class="custom-control-indicator"></span>
          <span class="custom-control-description sr-only">select</span>
        </label>
      </div>
    </div>
    <div class="flex-col flex-col-expand">
      <h4 class="list-group-title text-truncate">
        <a href="#1">Item 1</a>
      </h4>
      <p class="list-group-subtitle text-truncate">Description 1</p>
    </div>
  </li>
  <li class="list-group-item list-group-item-flex">
    <div class="flex-col">
      <div class="custom-control custom-checkbox">
        <label>
          <input class="custom-control-input" ref="input" type="checkbox">
          <span class="custom-control-indicator"></span>
          <span class="custom-control-description sr-only">select</span>
        </label>
      </div>
    </div>
    <div class="flex-col flex-col-expand">
      <h4 class="list-group-title text-truncate">
        <a href="#2">Item 2</a>
      </h4>
      <p class="list-group-subtitle text-truncate">Description 2</p>
    </div>
  </li>
  <li class="list-group-item list-group-item-flex">
    <div class="flex-col">
      <div class="custom-control custom-checkbox">
        <label>
          <input class="custom-control-input" ref="input" type="checkbox">
          <span class="custom-control-indicator"></span>
          <span class="custom-control-description sr-only">select</span>
        </label>
      </div>
    </div>
    <div class="flex-col flex-col-expand">
      <h4 class="list-group-title text-truncate">
        <a href="#3">Item 3</a>
      </h4>
      <p class="list-group-subtitle text-truncate">Description 3</p>
    </div>
  </li>
</ul>
`;

exports[`ClayListGroup should render a selectable ClayListGroup with header title 1`] = `
<ul class="list-group show-quick-actions-on-hover">
  <li class="list-group-header">
    <h3 class="list-group-header-title">Group Header 1</h3>
  </li>
  <li class="list-group-item list-group-item-flex">Item 1</li>
</ul>
`;

exports[`ClayListGroup should render the default markup 1`] = `<ul class="list-group show-quick-actions-on-hover"></ul>`;
